Reuse Cache in Computing Results, Implement #103

BenBals opened the pull request:
Note that if you don't use metacello for upgrading, you have to run `ALGItemProvider allInstances do: [:each | each cache: ALGCache new]` manually.

BenBals committed:
basic cache reuse

BenBals committed:
write tests for substring aware cache and fix a few errors

BenBals committed:
fix special case around empty string

BenBals committed:
remove dead code

BenBals committed:
test substring caching

BenBals committed:
remove code

BenBals committed:
remove more dead code

BenBals committed:
test item rank

BenBals committed:
remove dead code, code for upgrade

BenBals committed:
classify methods

BenBals committed:
fix linter errors

BenBals committed:
remove dead code

BenBals committed:
documentation test for selector usage

BenBals committed:
remove more dead code

BenBals committed:
special casing for non-regular filters

BenBals committed:
special casing need for the help filter

BenBals committed:
fix syntax error in baseline

BenBals committed:
ALGItemFilters overwrote selectors to signal all keywords the filter is applied to, but selectors is a method that also replys all methods of an object

BenBals committed:
fix fuzzy matching for help items

BenBals committed:
fix lints, refactor class variable on ALGItemFilter

BenBals committed:
move keywords on Comment filter to class side

sinnaj-r commented on packages/Algernon.package/ALGCache.class/instance/getAt..st:
```diff
@@ -1,6 +1,6 @@
 caching
 getAt: aCacheKey
```
I think we should overwrite the directory methods directly. Having both " at:" and "getAt:" at the same time is not really pretty and might confuse later developers.

BenBals replied:
Did that.

sinnaj-r commented on packages/Algernon.package/ALGCache.class/instance/putItems.at..st:
```diff
@@ -1,4 +1,4 @@
 caching
 putItems: anItemCollection at: aCacheKey
```
See comment for "getAt:"; I think this should be "at:put:" (or at least, "at:putItems:")

sinnaj-r commented on packages/Algernon.package/ALGCache.class/instance/clear.st:
```diff
@@ -1,4 +1,3 @@
 caching
 clear
```
Do we need this function ? As this ist just an "rename" for removeAll, we might want to call removeAll instead of clear.

BenBals replied:
This is because we need to ensure that old calls from `ALGInstallation instance cleanup` need work even if the new code has been loaded. That might happen during upgrade via Montecello. (I think, that stuff is black magic.)

sinnaj-r replied:
Okay, I see.

sinnaj-r reviewed (changes requested):
First of all the search is now (in comparison) ridiculously fast. This is an astonishing improvement! Great Job 👍 Perhaps we can also "afford" decrease the filteringThreshold, so that the search runs more often therefore showing the results quicker/more responsive as you type.

**But**: I have the Problem, that certain results are missing. E.g. when I search for Algernon I get no hits and when I search for "ALG" some results (The Tests, the ALGBackground-Class, etc.) are not in the results. I would guess that this is an Cache Reset/Init-Problem, do you have an Idea what I could try ?

BenBals committed:
rename ALGCache and refactor cache tests

BenBals committed:
fix typo in cache test setUp method

BenBals committed:
fix cache test expectations

BenBals committed:
fix cache test expectation (for real)

BenBals committed:
refactor ALGCache method names even more

BenBals committed:
write test reproducing substring matching cache bug

BenBals committed:
change caching to include all results

BenBals committed:
try to fix substring matching bug

This makes the matching *really* slow

BenBals committed:
fix slowness in matching items, does not produce the right behaivior however

BenBals committed:
fix partial caching issue

BenBals committed:
fix reference to old method

BenBals committed:
sort empty query in child list correctly

sinnaj-r commented on packages/Algernon.package/ALGItemFilter.class/instance/computeFuzzyItemsOutOf.given..st:
```diff
@@ -5,4 +5,4 @@ computeFuzzyItemsOutOf: anItemCollection given: aStringCollection
 	queryString := aStringCollection joinSeparatedBy: ' '.
 	anItemCollection do: [ :item | item fuzzyRank: (ALGFuzzyMatcher computeScore: item given: queryString)].
 	fuzzySortedItems := (anItemCollection select: [ :each | each fuzzyRank isNil not ]) sort: [ :a :b | (b fuzzyRank) < (a fuzzyRank)].
```
This line should be removed because it was moved to it's own method - correct ?

BenBals committed:
refactor method names in ItemProvider

BenBals committed:
remove unneeded filtering

BenBals committed:
remove line that was refactored into its own method

BenBals committed:
use set for class comment items

BenBals committed:
fix set operations in filter:given:

BenBals committed:
fix cache key generation for children

Previously we had issues because the search string was at the beginning of the cache key, so other queries would be errouneously found via substring search.

BenBals committed:
add second cache for better results on queries that have been performed previously

BenBals committed:
fix clearing of cache

BenBals committed:
fix cache tests
